src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java

```diff
@@ -15,15 +15,194 @@
+
+    return !errorHandlingStrategy.shouldRetry();
+  }
+
+  @Nullable
+  protected BasePage getNextPage() throws IOException {
+    // no more pages
+    if (nextPageUrl == null) {
+      return null;
+    }
+
+    // response being null, means it's the first page we are loading
+    Long delay = (response == null) ? 0L : config.getWaitTimeBetweenPages();
+    LOG.debug("Fetching '{}'", nextPageUrl);
+
+    try {
+      Awaitility
+        .await().with()
+        .pollInterval(pollInterval)
+        .pollDelay(delay, TimeUnit.MILLISECONDS)
+        .timeout(config.getMaxRetryDuration(), TimeUnit.SECONDS)
+        .until(this::visitPageAndCheckStatusCode);
+    } catch (ConditionTimeoutException ex) {
+      // Retries failed. We don't need to do anything here. This will be handled using httpStatusCode below.
+    }
+
+    HttpEntity reponseEntity = response.getEntity();
+    byte[] responseBytes = EntityUtils.toByteArray(reponseEntity);
+    String responseBody = new String(responseBytes, StandardCharsets.UTF_8);
+
+    HttpErrorHandlingStrategy postRetryStrategy = httpErrorHandler.getErrorHandlingStrategy(httpStatusCode)
+      .getAfterRetryStrategy();
+
+    String errorMessage = String.format("Fetching from url '%s' returned status code '%d' and body '%s'",
+                                        nextPageUrl, httpStatusCode, responseBody);
+
+    switch (postRetryStrategy) {
+      case SUCCESS:
+        break;
+      case FAIL:
+        throw new IllegalStateException(errorMessage);
+      case SKIP:
+      case SEND_TO_ERROR:
+        if (!this.supportsSkippingPages()) {
+          throw new IllegalStateException(String.format(
+            "Pagination type '%s', does not support 'skip' and 'send to error' error handling.",
+            config.getPaginationType()));
+        }
+        LOG.warn(errorMessage);
+        // this will be handled by PageFactory. Here no handling is needed.
+        break;
+      default:
+        throw new IllegalArgumentException(String.format("Unexpected http error handling: '%s'", postRetryStrategy));
+    }
+
+    if (config.getFormat().equals(PageFormat.BLOB)) {
+      responseBody = new String(Base64.encodeBase64(responseBytes), StandardCharsets.UTF_8);
+    }
+
+    BasePage page = createPageInstance(config, responseBody, postRetryStrategy);
+    nextPageUrl = getNextPageUrl(responseBody, response, page);
+
+    LOG.debug("Next Page Url is '{}'", nextPageUrl);
+    //LOG.info("Schema is: {}", config.getSchemaString());
+
+    return page;
+  }
+
+  /**
+   * @return true if page still has elements to iterate. Otherwise it will load next page.
+   * False if no more pages to load or the page loaded has no elements.
+   */
+  protected boolean ensurePageIterable() {
+    try {
+      if (page == null || !page.hasNext()) {
+        page = getNextPage();
+      }
+
+      return page != null && page.hasNext();
```

Comment:
Empty page (with no elements) in pagination APIs usually means that you should stop paginating otherwise you'll get into endless loop.

Example API:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=80

You can go to page 100 200 500 1000 which are empty and it's endless. All with no elements.
Page 10.000:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=10000

Reply:
that's why I don't think we need any change here.

Reply:
I see, makes sense